client_it_test.go

```diff
@@ -553,3 +558,61 @@ func TestClientFailover_EECluster_Reconnection(t *testing.T) {
 		t.Fatal(err)
 	}
 }
+
+func highlight(t *testing.T, format string, args ...interface{}) {
+	log.Printf("\n===\n%s\n===", fmt.Sprintf(format, args...))
+}
+
+func TestClientFixConnection(t *testing.T) {
+	// This test removes the member that corresponds to the connections which receives membership state changes.
+	// Once that connection is closed, another connection should be randomly selected to receive membership state changes.
+	// A new member is added to confirm that is the case.
+	const memberCount = 3
+	addedCount := int64(0)
+	ctx := context.Background()
+	id := idGen.NextID()
+	clusterName := fmt.Sprintf("600-cluster-%d", id)
+	log.Println("Cluster name:", clusterName)
+	port := 20701 + id*10
+	cls := it.StartNewClusterWithOptions(clusterName, int(port), memberCount)
+	defer cls.Shutdown()
+	config := hz.Config{}
+	config.Cluster.Network.SetAddresses(fmt.Sprintf("localhost:%d", port+1))
+	config.Cluster.Name = clusterName
+	config.AddMembershipListener(func(event cluster.MembershipStateChanged) {
+		highlight(t, "%s member: %s", event.State.String(), event.Member.UUID)
+		if event.State == cluster.MembershipStateAdded {
+			atomic.AddInt64(&addedCount, 1)
+		}
+	})
+	if it.TraceLoggingEnabled() {
+		config.Logger.Level = logger.TraceLevel
+	}
+	client, err := hz.StartNewClientWithConfig(ctx, config)
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer client.Shutdown(ctx)
+	// terminate the member that corresponds to the connection which receives cluster membership updates
+	mUUID := cls.MemberUUIDs[1]
+	highlight(t, "Terminated member: %s", mUUID)
+	ok, err := cls.RC.TerminateMember(ctx, cls.ClusterID, mUUID)
+	if err != nil {
+		log.Fatal(err)
+	}
+	if !ok {
+		log.Fatalf("could not terminate member: %s", err.Error())
+	}
+	m, err := cls.RC.StartMember(ctx, cls.ClusterID)
+	if err != nil {
+		log.Fatal(err)
+	}
+	highlight(t, "Started member: %s", m.UUID)
+	time.Sleep(30 * time.Second)
+	assert.Equal(t, int64(memberCount+1), atomic.LoadInt64(&addedCount))
```

Comment:
Can we implement something like assertTrueEventually that we have on the java to avoid sleep statements in the tests. https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/test/java/com/hazelcast/test/HazelcastTestSupport.java#L1251

Reply:
We can something like timed waitgroups. I've created a ticket for that: https://github.com/hazelcast/hazelcast-go-client/issues/621